File: diary-app-backend/DiaryApp.Data/Services/ListService.cs
using System.Linq;
using System.Threading.Tasks;
using DiaryApp.Core;
using DiaryApp.Core.Models.Lists;

namespace DiaryApp.Data.Services
{
    public abstract class ListService<TList, TItem> : CrudService<TList>, IListService<TList, TItem> 
        where TList : ListBase<TItem> 
        where TItem : ListItemBase
    {
        private readonly CrudService<TItem> itemsService;
        public ListService(ApplicationContext context) : base(context)
        {
            itemsService = new CrudService<TItem>(context);
        }

        public async Task AddItem(TItem item, int ownerId)
        {
            var eventList = await GetById(ownerId);
            if (eventList != null)
            {
                eventList.Items.Add(item);
                await context.SaveChangesAsync();
            }
        }

        public async Task DeleteItem(int itemID)
        {
            await itemsService.Delete(itemID);
        }

        public async Task<TItem> GetItemByID(int itemID)
        {
            return await itemsService.GetById(itemID);
        }

        public IQueryable<TItem> GetItems()
        {
            return itemsService.GetAll();
        }

        public async Task UpdateItem(TItem item)
        {
            await itemsService.Update(item);
        }
    }
}

File: diary-app-frontend/src/services/listService.ts
import axios from "../axios/axios";
import { IList, IListItem } from "../models";

export const getListService = <T extends IList<TItem>, TItem extends IListItem>(
	apiUrl: string
) => {
	const createList = (list: T) => {
		return axios.post<number>(apiUrl, list).then((res) => res.data);
	};

	const updateList = (list: T) => {
		return axios.put(apiUrl, list).then((res) => res.data);
	};

	const addItem = (item: TItem) => {
		return axios.post<number>(`${apiUrl}items`, item).then((res) => res.data);
	};

	const updateItem = (item: TItem) => {
		return axios.put(`${apiUrl}items`, item).then((res) => res.data);
	};

	const deleteItem = (itemID: number) => {
		return axios.delete(`${apiUrl}items/${itemID}`).then((res) => res.data);
	};

	const deleteList = (listID: number) => {
		return axios.delete(apiUrl + listID).then((res) => res.data);
	};

	return {
		createList,
		updateList,
		deleteList,
		addItem,
		updateItem,
		deleteItem,
	};
};
